kong/dns/utils.lua

```diff
@@ -0,0 +1,274 @@
+  if not lines then
+    log(NOTICE, "Invalid hosts file: ", err)
+    return DEFAULT_HOSTS
+  end
+
+  local hosts = {}
+
+  for _, line in ipairs(lines) do
+    -- Remove leading/trailing whitespaces and split by whitespace
+    local parts = {}
+    for part in line:gmatch("%S+") do
+      if part:sub(1, 1) == '#' then
+        break
+      end
+      table_insert(parts, part:lower())
+    end
+
+    -- Check if the line contains an IP address followed by hostnames
+    if #parts >= 2 then
+      local ip, _, family = _M.parse_hostname(parts[1])
+      if family ~= "name" then    -- ipv4/ipv6
+        for i = 2, #parts do
+          local host = parts[i]
+          local v = hosts[host]
+          if not v then
+            v = {}
+            hosts[host] = v
+          end
+          v[family] = v[family] or ip -- prefer to use the first ip
+        end
+      end
+    end
+  end
+
+  if not hosts.localhost then
+    hosts.localhost = LOCALHOST
+  end
+
+  return hosts
+end
+
+
+-- TODO: need to rewrite it instead of calling parseResolvConf from the old library
+function _M.parse_resolv_conf(path, enable_ipv6)
+  local resolv, err = utils.parseResolvConf(path or DEFAULT_RESOLV_CONF)
+  if not resolv then
+    return nil, err
+  end
+
+  resolv = utils.applyEnv(resolv)
+  resolv.options = resolv.options or {}
+  resolv.ndots = resolv.options.ndots or 1
+  resolv.search = resolv.search or (resolv.domain and { resolv.domain })
+
+  -- check if timeout is 0s
+  if resolv.options.timeout and resolv.options.timeout <= 0 then
+    log(NOTICE, "A non-positive timeout of ", resolv.options.timeout,
+                "s is configured in resolv.conf. Setting it to 2000ms.")
+    resolv.options.timeout = 2000 -- 2000ms is lua-resty-dns default
+  end
+
+  -- remove special domain like "."
+  if resolv.search then
+    for i = #resolv.search, 1, -1 do
+      if resolv.search[i] == "." then
+        table_remove(resolv.search, i)
+      end
+    end
+  end
+
+  -- nameservers
+  if resolv.nameserver then
+    local nameservers = {}
+
+    for _, address in ipairs(resolv.nameserver) do
+      local ip, port, t = utils.parseHostname(address)
+      if t == "ipv4" or
+        (t == "ipv6" and not ip:find([[%]], nil, true) and enable_ipv6)
+      then
+        table_insert(nameservers, port and { ip, port } or ip)
+      end
+    end
+
+    resolv.nameservers = nameservers
+  end
+  return resolv
+end
+
+
+function _M.is_fqdn(name, ndots)
+  if name:sub(-1) == "." then
+    return true
+  end
+  local _, dot_count = name:gsub("%.", "")
+  return (dot_count >= ndots)
+end
+
+
+-- construct names from resolv options: search, ndots and domain
+function _M.search_names(name, resolv, hosts)
+  if not resolv.search or _M.is_fqdn(name, resolv.ndots) then
+    return { name }
+  end
+
+  local names = {}
+  for _, suffix in ipairs(resolv.search) do
+    table_insert(names, name .. "." .. suffix)
+  end
+
+  if hosts and hosts[name] then
```

Comment:
Should have a comment to explain this logic.

Reply:
https://github.com/Kong/kong/pull/12305/commits/b87cc14ab796307cf88d07c7aed08475a9bb0497 done